File: Codes/MATLAB Codes/matrix_vector_example_convolutional_codes.m
%   Finding the worst case error percentage for our proposed scheme
%   Having a random matrix A of size (r,t) and a vector x of length t.
%   We have n workers, s stragglers.
%   Storage fraction gamma > 1/k ; where k = n - s; and find Delta.
%   Matrix A is divided into Delta block columns, where k divides Delta.
%   One can vary SNR to check the scenario at different noise levels.
%   Set random = 1 for an upper bound for the condition number.
%   Set worst_case = 1 to find the error in the worst_case scenario.
%   worst_condition_number is the maximum condition number in this scheme.
%   Set dist = 'rand' if you want the coefficients from standard normal distribution.
%   One can increase no_trials, which can help to find a better condition number.  

clc
close all
clear

n = 10;
s = 3;                                     %% set straggler number, s > 1
k = n - s;
gamma = 1/5;                               %% needs to be greater than 1/k
Delta = round((s-1)*(k-1)/(gamma - 1/k));
while(rem(Delta,k)~=0)
    Delta = Delta + 1;
end
q = Delta/k;

r = 6300;                                      %% Needs to be a multiple of Delta
t = 2000;
A = randn(t,r);
x = randn(t,1);
b = A'*x;
SNR = 50;
random = 1;                                    %% set 1 to choose random coefficients.
peeling = 0;
if random == 0
    peeling = 1;                               %% Peeling decoder for all 1's case
end
no_trials = 20;                                %% number of trials, if random = 1
dist = 'rand';                                 %% distribution, 'rand' or 'unif'
worst_case = 1;                                %% set 1 to find the worst case error.
if worst_case ~= 1
    all_workers = randperm(n);              
    active_workers = all_workers(1:k);         %% any k out of 1 to n
    active_workers = sort(active_workers);
end

if random == 1
    [best_matrix,predicted_upper_bound] = matrix_vector_best_mat(n,s,no_trials,dist);
else
    best_matrix = ones(k,s);
end
full_mat = [eye(k) best_matrix];

if worst_case == 1
    for i=1:k
        P = zeros(q,Delta);
        P(1:q,(i-1)*q+1:(i)*q)=eye(q);
        W{i}=P;                                     %% Matrix for Message Workers
    end
    
    zer = 2*Delta;
    for i=k+1:n
        W{i} = zeros(zer,Delta);
        P = [];
        for j=1:k
            P = shiftrow(full_mat(j,i)*W{j},(j-1)*(i-k-1),zer);
            W{i} = W{i} + P;                        %% Matrix for Parity Workers
        end
        W{i}(~any(W{i},2),:)=[];
    end
    
    workers = 1:n;
    Choice_of_workers = combnk(workers,k);          %% Different sets of workers
    [total_no_choices,~] = size(Choice_of_workers);
    cond_no = zeros(total_no_choices,1);
    
    for i = 1:total_no_choices
        Coding_matrix = [];
        for j = 1:k
            Coding_matrix = [Coding_matrix ; W{Choice_of_workers(i,j)}];
        end
        cond_no(i) = cond(Coding_matrix);
    end
    
    worst_condition_number = max(cond_no);
    pos = find(cond_no==worst_condition_number);
    active_workers = Choice_of_workers(pos,:);
    if random ==1
        fprintf('\n'); 
        disp(['The upper bound of the condition number is ', num2str(predicted_upper_bound),'.']);
    end
    fprintf('\n'); 
    disp(['The worst case condition number is ', num2str(worst_condition_number),'.']);
    fprintf('\n'); 
    disp(['The worst case includes workers ', num2str(active_workers),'.']);
end


%%%%% Encoding Starts in the Master Node  %%%%%%

for i=1:k
    W1{i} = A(:,(i-1)*r/k+1:i*r/k);             %% Diving Matrix A in k blocks
end

c = r/Delta;
for i=1:k
    for j =1:q
        W2{i}{j} = W1{i}(:,(j-1)*c+1:j*c);      %% Message Workers' Assignments
    end
end

for i = k+1:n
    len = q + (k-1)*(i-k-1);
    for j =1:len
        sumA = zeros(t,c);
        for rr = 1:k
            if j >= (i-k-1)*(rr-1)+1 && j <= (i-k-1)*(rr-1)+q
                sumA  = sumA + W2{rr}{j-(i-k-1)*(rr-1)}* full_mat(rr,i);
            end
        end
        W2{i}{j} = sumA;                         %% Parity Workers' Assignments
    end
end

%%%%% Computation starts in the worker nodes  %%%%%%

for i = 1:n
    [~,len]=size(W2{i});
    for j=1:len
        W3{i}{j} = awgn(W2{i}{j}'*x,SNR);        %% Worker Computation and Noise
    end
end

for i=1:k
    Coding_matrix = zeros(q,Delta);
    Coding_matrix(1:q,(i-1)*q+1:(i)*q)=eye(q);
    W{i}=Coding_matrix;
end

zer = 2*q*k;
for i=k+1:n
    W{i} = zeros(zer,Delta);
    P = [];
    for j=1:k
        P = shiftrow(full_mat(j,i)*W{j},(j-1)*(i-k-1),zer);
        W{i} = W{i} + P;
    end
    W{i}(~any(W{i},2),:)=[];
end

%%%%%%%% Decoding Starts in Master Node %%%%%

tic

amw = active_workers(active_workers<=k);
apw = active_workers(active_workers>k);

res = zeros(c*q,k);
for i = 1:length(amw)
    for j = 1:q
        res((j-1)*c+1:j*c,amw(i)) = W3{amw(i)}{j};      %% getting the values from the systematic part
    end
end

for i = k+1:n
    if ~ismember(i,apw), continue; end
    len = q + (k-1)*(i-k-1);
    for j =1:len
        sumA = zeros(c,1);
        for rr = 1:k
            if ~ismember(rr,amw), continue; end
            if j >= (i-k-1)*(rr-1)+1 && j <= (i-k-1)*(rr-1)+q
                sumA  = sumA + W3{rr}{j-(i-k-1)*(rr-1)}* full_mat(rr,i);
            end
        end
        W3{i}{j} = W3{i}{j} - sumA;
    end
end

Coding_matrix = [];
for j = 1:length(apw)
    Coding_matrix = [Coding_matrix ; W{active_workers(length(amw)+j)}];
end

[xx,yy] = size(Coding_matrix);
if xx>0
    for i =1:k
        if ismember(i,amw)
            Coding_matrix(:,(i-1)*q+1:i*q)=0;
        end
    end
end
output_worker = [];
for i=1:length(apw)
    [~,len]=size(W3{active_workers(length(amw)+i)});
    for j=1:len
        output_worker = [output_worker;W3{active_workers(length(amw)+i)}{j}'];
    end
end
fprintf('\n');

if peeling == 0                                       %% LS Decoding for random convolutional coding
    Coding_matrix(:,~any(Coding_matrix,1))=[];
    res2 = sparse(Coding_matrix)\output_worker;       %% results from the parity parts using LS
    res_trans = res2';
    coded_res = res_trans(:);
    
    count = 0;
    for i = 1:k
        if ~ismember(i,amw)
            count = count + 1;
            res(:,i) = coded_res((count-1)*c*q+1:count*c*q);
        end
    end
    
else                            %% Peeling Decoder for all 1's
    res2 = (reshape(res(:),[c,Delta]))';
    AA = Coding_matrix;
    BB = output_worker;
    zer_rows = find(all(AA==0,2));
    AA(zer_rows,:)=[];
    BB(zer_rows,:)=[];
    
    while any(AA(:))
        [~,imp_rows] = unique(AA,'rows');       %% removing the unique rows
        imp_rows = sort(imp_rows);
        AA = AA(imp_rows,:);
        BB = BB(imp_rows,:);
        
        [u,v] = size(AA);
        ind1 = sum(AA~=0,2);                    %% Finding the rows with single unknown
        ind2 = find(ind1==1);
        [aa,bb] = find(AA(ind2,:)~=0);
        [aa,ij] = sort(aa);
        bb = bb(ij);
        res2(bb,:) = BB(ind2,:);                %% Recovering unknowns
        [ee,ff] = find(AA(:,bb)~=0);
        for ii = 1:length(ee)
            BB(ee(ii),:) = BB(ee(ii),:)- res2(bb(ff(ii)),:);
        end
        
        AA(ee,bb(ff))=0;
        AA(ind2,:)=[];                          %% Removing the already used rows
        BB(ind2,:)=[];
    end
    res = res2';
end

final_result = res(:);
decoding_time = toc;
disp(['The decoding time is ', num2str(decoding_time),'seconds.']);
err_ls = 100*(norm(final_result-b,'fro')/norm(b,'fro'))^2;
fprintf('\n');
disp(['The error is ', num2str(err_ls),' % .']);
fprintf('\n');

File: Codes/Python Codes/matrix_vector_example_convolutional_codes.py
"""
Having a random matrix A of size (r,t) and a vector x of length t.
Finding the worst case error percentage for our proposed scheme to find A'x
We have n workers, s stragglers.
Storage fraction gamma > 1/k ; where k = n - s; and find Delta.
Matrix A is divided into Delta block columns, where k divides Delta.
One can vary SNR to check the scenario at different noise levels.
Set random = 1 for an upper bound for the condition number.
Set worst_case = 1 to find the error in the worst_case scenario.
Worst_condition_number is the maximum condition number in this scheme.
One can increase no_trials, which can help to find a better condition number.  
"""

import numpy as np
from scipy import linalg
import itertools as it
import time
from scipy.sparse import csr_matrix


def shiftrow(A,r,t):   
    (a,b) = np.shape(A) ;
    B = np.zeros((t,b),dtype = float) ;
    B[r:a+r,:] = A ;
    return B;

def matrix_vector_best_mat(n,s,no_trials):
    samples_in_omega = 400;
    k = n - s ;
    message = np.identity(k,dtype = float);
    condition_number = np.zeros(no_trials,dtype = float);
    parity = {} ;
    workers = list(range(n));
    Choice_of_workers = list(it.combinations(workers,k));
    size_total = np.shape(Choice_of_workers);            
    total_no_choices = size_total[0];
    min_eigenvalue = np.zeros(total_no_choices*samples_in_omega);
    max_eigenvalue = np.zeros(total_no_choices*samples_in_omega);
    
    for t in range (0,no_trials):
        parity[t] = np.random.rand(k,s);
        ind = 0;
        min_eigenvalue = np.zeros(total_no_choices*samples_in_omega);
        max_eigenvalue = np.zeros(total_no_choices*samples_in_omega);
        w = np.zeros(2*samples_in_omega,dtype = float);
        for i in range (0,samples_in_omega):
            w[i] = -np.pi + i*2*np.pi/samples_in_omega;
        zz = samples_in_omega;
        for z in range (0,zz):
            imag = 1j;
            omega = np.zeros((k,s),dtype = complex);
            for i in range (0,s):
                omega[:,i] = np.power(np.exp(-imag*w[z])**i,list(range(k)))
            Generator_mat = np.concatenate((message, np.multiply(parity[t],omega)), axis = 1)
            for i in range (0,total_no_choices):
                Coding_matrix = [];
                kk = list(Choice_of_workers[i]);
                Coding_matrix = Generator_mat[:,kk];
                Coding_matrixT = np.transpose(Coding_matrix);
                D = np.matmul(np.conjugate(Coding_matrixT),Coding_matrix);
                eigenvalues = linalg.eigvals(D);
                eigenvalues = np.real(eigenvalues);
                min_eigenvalue[ind] = np.min(eigenvalues);
                max_eigenvalue[ind] = np.max(eigenvalues);
                ind = ind + 1;
        condition_number[t] = np.sqrt(np.max(max_eigenvalue)/np.min(min_eigenvalue)) 
    best_cond_min = np.min(condition_number);
    position =   np.argmin(condition_number);
    R = parity[position];
    return R,best_cond_min        
            

n = 10 ;
s = 3 ;                                             ## set straggler number, s > 1
k = n - s ;
gamma = 1/5;                                        ## needs to be greater than 1/k
SNR = 80; 
Delta = int(np.round((s-1)*(k-1)/(gamma-1/k))); 
while Delta % k != 0:
    Delta = Delta+1 ;
q = int(Delta/k) ;                                 ## Needs to be a multiple of Delta
print('\nThe value of Delta is %s' % Delta)
r = 6300 ;
t = 1000 ;
mu = 0;
sigma = 1;
A = np.random.normal(mu, sigma, [t,r]);
x = np.random.normal(mu, sigma, [t,1]);
b = np.matmul(np.transpose(A),x);
random = 1;                                        ## set 1 to choose random coefficients.
no_trials = 25;                                     ## number of trials, if random = 1
worst_case = 1;                                    ## set 1 to find the worst case error

peeling = 0;
if random == 0:
    peeling = 1;                                  ## Peeling decoder for all 1's case
    

if worst_case !=1:
    all_workers = np.random.permutation(n);
    active_workers = all_workers[0:k];
    active_workers.sort() ;

if random == 1:
    (best_matrix,predicted_upper_bound) = matrix_vector_best_mat(n,s,no_trials);
else:
    best_matrix = np.ones((k,s),dtype = float);

full_mat = np.concatenate((np.identity(k,dtype = float),best_matrix), axis = 1) ;
W = {} ;
if worst_case == 1:
    for i in range (0,k):
        P = np.zeros((q,Delta),dtype=float);
        P[0:q,i*q:(i+1)*q] = np.identity(q,dtype = float);
        W[i] = P;                               ## Matrix for Message Workers

    zer = 2*Delta ;
    for i in range (k,n):
        W[i] = np.zeros((zer,Delta),dtype=float);
        P = [];
        for j in range (0,k):
            P = shiftrow(full_mat[j,i]*W[j],j*(i-k),zer);
            W[i] = W[i] + P;                    ## Matrix for Parity Workers
        inter = W[i];
        W[i] = inter[~np.all(inter == 0, axis=1)];
    workers = list(range(n));
    Choice_of_workers = list(it.combinations(workers,k));          ## Different sets of workers
    size_total = np.shape(Choice_of_workers);            
    total_no_choices = size_total[0];
    cond_no = np.zeros(total_no_choices,dtype = float);    
    for i in range (0, total_no_choices):
        for j in range (0,k):
            kk = list(Choice_of_workers[i]);
            if j==0:
                Coding_matrix = W[kk[j]];
            else:
                Coding_matrix = np.concatenate((Coding_matrix, W[kk[j]]), axis = 0);
        cond_no[i] = np.linalg.cond(Coding_matrix);    
    worst_condition_number = np.max(cond_no);
    pos =   np.argmax(cond_no);
    active_workers = list(Choice_of_workers[pos]);       
    if random == 1:
        print('\nUpper Bound of the condition number is %s' % predicted_upper_bound);
    print('\nWorst case condition number is %s' % worst_condition_number)
    print('\nActive workers are %s' %active_workers)          

W1 = {} ;
for i in range (0,k):
    W1[i] = A[:,int(i*r/k):int((i+1)*r/k)] ;  
del A
W2 = {} ;
c = int(r/Delta);
for i in range (0,k):
    for j in range (0,q):
        W2[i,j] = W1[i][:,j*c:(j+1)*c] ;                ## Message Workers' Assignments
del W1
for i in range (k,n):
    len = q + (k-1)*(i-k);
    for j in range (0,len):
        sumA = np.zeros((t,c),dtype = float);
        for rr in range (0,k):
            if j >= (i-k)*rr and j <= (i-k)*rr+q-1:
                sumA = sumA + W2[rr,j-(i-k)*rr]*full_mat[rr,i];
                
        W2[i,j] = sumA;                                 ##  Parity Workers' Assignments
for i in range (0,n):
    leng = q if i < k else q + (k-1)*(i-k) ;
Wsend = {};
for j in range (0,leng):
	 Wsend[j] = W2[i,j];
W3 = {} ;        
for i in range (0,n):
    len = q if i < k else q + (k-1)*(i-k) ;
    for j in range (0,len):
        W3[i,j] = np.matmul(np.transpose(W2[i,j]),x);
        sig = W3[i,j];
        sig_avg_db = 10*np.log10(np.mean(sig ** 2));
        noise_avg_db = sig_avg_db - SNR;
        var = 10 ** (noise_avg_db / 10);
        W3[i,j] = W3[i,j] + np.random.normal(0,np.sqrt(var),np.shape(W3[i,j]));
        
W = {};
for i in range (0,k):
    Coding_matrix = np.zeros((q,Delta),dtype = float);
    Coding_matrix[0:q,i*q:(i+1)*q] = np.identity(q,dtype = float);
    W[i] = Coding_matrix ; 
    
zer = 2*q*k;
for i in range (k,n):
    W[i] = np.zeros((zer,Delta));
    P = [];
    for j in range (0,k):
        P = shiftrow(full_mat[j,i]*W[j],j*(i-k),zer);
        W[i] = W[i] + P ;
    inter = W[i];
    W[i] = inter[~np.all(inter == 0, axis=1)]   


start_time = time.time()
ss = np.array(active_workers)
amw = ss[ss<k];
apw = ss[ss>=k];

res = np.zeros((c*q,k),dtype = float);
for i in range(0,np.size(amw)):
    for j in range(0,q):
        dd = W3[amw[i],j];
        res[j*c:(j+1)*c,amw[i]]  = dd.ravel()           ## getting the values from the systematic part

for i in range(k,n):
    if i in apw:
        leng = q + (k-1)*(i-k);
        for j in range (0,leng):
            sumA = np.zeros((c,1),dtype = float);
            for rr in range(0,k):
                if rr in amw:
                    if j >= (i-k)*rr and j <= (i-k)*rr+q-1:
                        sumA = sumA + full_mat[rr,i]*W3[rr,j-(i-k)*rr];
            W3[i,j] = W3[i,j]-sumA

           
if np.size(apw)>0:
    Coding_matrix = W[active_workers[np.size(amw)]];
    for j in range (1,np.size(apw)):
        Coding_matrix = np.concatenate((Coding_matrix, W[active_workers[np.size(amw)+j]]), axis = 0) ;

[xx,yy]=np.shape(Coding_matrix)
if xx>0:
    for i in range (0,k):
        if i in amw:
            Coding_matrix[:,i*q:(i+1)*q] = 0;

for i in range (0,np.size(apw)):
    leng = q + (k-1)*(active_workers[np.size(amw)+i]-k) ;
    for j in range (0,leng):
        if i==0 and j==0:
            output_worker = np.transpose(W3[active_workers[np.size(amw)],0]);
        else:
            output_worker = np.concatenate((output_worker, np.transpose(W3[active_workers[np.size(amw)+i],j])), axis = 0) ;       
          
if peeling == 0:                               ## LS Decoding for random convolutional coding
    inter = np.transpose(Coding_matrix);
    Coding_matrix = np.transpose(inter[~np.all(inter == 0, axis=1)])  
    CC = np.linalg.inv(np.matmul(np.transpose(Coding_matrix),Coding_matrix))
    res_p = np.matmul(CC,np.matmul(np.transpose(Coding_matrix),output_worker))
    
    atw = list(range(k));
    aow = [i for i in atw if i not in amw];
    res_p = np.reshape(res_p, (q*c*np.size(aow), 1));

    for j in range(0,np.size(apw)):
        res[:,aow[j]] = res_p[j*c*q:(j+1)*c*q].ravel();

else:                                   ## Peeling Decoder for all 1's
    AA = csr_matrix(Coding_matrix);
    BB = output_worker;
    res2 = np.transpose(res);
    res2 = (np.reshape(res2, (Delta,c)));
    while csr_matrix.count_nonzero(AA)>0:
        ind1 = AA.sum(axis=1);                      ## Finding the rows with single unknown
        ind3 = np.where(ind1==1)
        ind2 = ind3[0];
        (aa,bb) = np.nonzero(AA[ind2,:])
        ij = np.argsort(aa)
        bb = bb[ij];
        (_, imp_rows) = np.unique(bb, axis=0,return_index=True);  
        bb = bb[imp_rows];
        res2[bb,:] = BB[ind2[imp_rows],:];                    ## Recovering unknowns
        (ee,ff) = np.nonzero(AA[:,bb])        
        for ii in range(0,np.size(ee)):
            BB[ee[ii],:] = BB[ee[ii],:] - res2[bb[ff[ii]],:];
            
        AA[ee,bb[ff]] = 0;

    res = np.transpose(res2);

res = np.transpose(res);
final_result = np.reshape(res, (r, 1));
end_time = time.time();

print('\nTime needed is %s' % (end_time - start_time))
err_ls = 100*np.square(np.linalg.norm(final_result - b,'fro')/np.linalg.norm(b,'fro'));
print('\nError percentage is %s' % err_ls)
print('\n')
